inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/TableMetricStatementExecutor.java

```diff
@@ -0,0 +1,172 @@
+        this.stmtFactory = checkNotNull(stmtFactory);
+        this.converter = checkNotNull(converter);
+        this.batch = new ArrayList<>();
+        this.dirtySinkHelper = dirtySinkHelper;
+        this.sinkMetricData = sinkMetricData;
+    }
+
+    /**
+     * parses an SQL exception message, and returns dirty records
+     *
+     * @param e the exception
+     * @return a string that can be compared with the return value of other parseRecord calls
+     */
+    public static void parseRecord(Exception e, List<Integer> answer) {
+        final Pattern pattern = Pattern.compile("Batch entry (\\d+) ");
+        Matcher matcher = pattern.matcher(e.getMessage());
+        if (matcher.find()) {
+            answer.add(Integer.parseInt(matcher.group(1)));
+        }
+        // if e is sql exciption, identify all dirty data, else identify only one dirty data.
+        if (e instanceof SQLException) {
+            SQLException next = ((SQLException) e).getNextException();
+            if (next != null) {
+                parseRecord(next, answer);
+            }
+        }
+    }
+
+    @Override
+    public void prepareStatements(Connection connection) throws SQLException {
+        st = stmtFactory.createStatement(connection);
+    }
+
+    @Override
+    public void addToBatch(RowData record) throws SQLException {
+        batch.add(record);
+        converter.toExternal(record, st);
+        st.addBatch();
+    }
+
+    @Override
+    public void executeBatch() throws SQLException {
+        try {
+            st.executeBatch();
+            addMetrics();
+        } catch (SQLException e) {
+            if (counter.incrementAndGet() == 3) {
+                // parse record from error and handle
+                List<Integer> dirtyRecords = new ArrayList<>();
+                parseRecord(e, dirtyRecords);
+                handleDirty(dirtyRecords);
+                counter.set(0);
+            } else {
+                throw new SQLException(e);
+            }
+        }
+    }
+
+    private void addMetrics() {
+        Set<RowData> set = new HashSet<>();
+        int rowCount = 0;
+        long rowSize = 0;
+        for (RowData record : batch) {
+            if (!set.contains(record)) {
+                set.add(record);
+                rowCount++;
+                rowSize += record.toString().getBytes(StandardCharsets.UTF_8).length;
+            }
+        }
+        sinkMetricData.invoke(rowCount, rowSize * 8);
+        batch.clear();
+    }
+
+    private void handleDirty(List<Integer> dirtyRecords) throws SQLException {
+        Long rowCount = Long.valueOf(dirtyRecords.get(0));
+        Long rowSize = 0L;
+        List<RowData> toClear = new ArrayList<>();
+        // find statements
+        for (int i = 0; i < dirtyRecords.get(0); i++) {
+            toClear.add(batch.get(i));
+            rowSize += batch.get(i).toString().getBytes(StandardCharsets.UTF_8).length;
+        }
+        sinkMetricData.invoke(rowCount, rowSize);
```

Comment:
why to sinkMetricData.invoke(rowCount, rowSize);

Reply:
this should be like this. the data tends to be clean-dirty-clean,and the identifier here are the position(or first position) of dirty data. here, I need to add the number and size of all clean data, and then invoke those clean instead of the entire batch.

Reply:
@Yizhou-Yang dirtyRecords.get(0) is normal data?